packages/website/blog/2024-03-25-changes-to-consistent-type-imports-with-decorators.md

```diff
@@ -0,0 +1,101 @@
+
+<!-- truncate -->
+
+## Experimental Decorator Metadata
+
+The `experimentalDecorators` compiler option (referred to as "legacy decorators" from here on) turns on support for an old version of the decorator proposal that was never standardized. Syntactically this proposal isn't too dissimilar to the current [Stage 3 proposal](https://github.com/tc39/proposal-decorators), however they differ in one major way - the use of [metadata reflection](https://rbuckton.github.io/reflect-metadata/) when `emitDecoratorMetadata` is turned on.
+
+When using legacy decorators with decorator metadata and a class is annotated with decorators TypeScript will emit runtime metadata for the class that captures the property types, method parameter types, and method return types. This metadata is incredibly powerful and provides a bridge between the types (which are elided at compile time) and the runtime code. However this runtime code is generated at a cost as it is derived using type information - meaning that the runtime code emitted changes based on the cross-file type information that TypeScript has computed.
+
+To illustrate what this means consider the following snippet:
+
+```ts
+import Foo from 'foo';
+import decorator from 'decorator';
+
+class Clazz {
+  @decorator
+  method(arg: Foo) {}
+}
+```
+
+If the imported name `Foo` resolves to...
+
+- a type then TS will emit metadata that annotates `arg` with one of `Function`, `Object`, `String`, `Number`, or `Boolean` depending on what that type resolves to.
+- an enum then TS will emit metadata that annotates `arg` with one of `String`, `Number`, or `Object` depending on the type of the enum's members.
+- a class declaration:
+  - and the import **_is NOT_** annotated as `import type` then TS will emit metadata that annotates the type of `arg` as `Foo`.
+  - and the import **_IS_** annotated as `import type` then TS will emit metadata that annotates the type of `arg` as `Function`.
+
+## `consistent-type-imports` caused runtime breakage
+
+The important piece is that last dot point - the handling of imported names that resolve to class declarations. If the import is not annotated as `import type` then TS emits a runtime reference to the imported name. This runtime reference is implicit and requires type information to derive - you cannot derive its existence purely based on single-file AST analysis.
+
+The `consistent-type-imports` rule was introduced to allow users to enforce that any imported names are annotated as `import type` if they are not used in a value location. How the rule makes this decision is based on single-file AST analysis; it scans the code using a technique called scope analysis so that it can find all references to the imported names and determine if each reference is a value reference.
```

Comment:
Devs reading this might not know what an AST is. And they don't need to in order to understand that it's a one-file-at-a-time thing. This is a good place to suggest they learn it though!

```suggestion
The `consistent-type-imports` rule was introduced to allow users to enforce that any imported names are annotated as `import type` if they are not used in a value location. How the rule makes this decision is based solely on the single file it's looking at.
It scans the code using a technique called "scope analysis" so that it can find all references to the imported names and determine if each reference is a value reference.

:::tip
See [ASTs and typescript-eslint](../asts-and-typescript-eslint) to understand how rules look at the syntax of files.
:::
```

Reply:
Aside: I found it a little confusing to read through this. I think the point that `consistent-type-imports` tries to be an untyped rule tackling a typed problem isn't super clear immediately. This was a vague attempt at rewriting some of it:

> Where the `consistent-type-imports` rule used to trip up is around TypeScript's code emit including a runtime reference to any import not annotated as `import type`. That runtime reference is implicit and requires type information to derive. TypeScript needs to understand the type of the import to emit the metadata. But, `consistent-type-imports` was an untyped lint rule. It didn't have access to type information to know whether an import used as a decorator could be switched from `import` to `import type`!